### PR TITLE
Web

### PR DESCRIPTION
**Pull request by filip373**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,6 +1,9 @@
 source "https://rubygems.org"
 
-group :test do
-  gem 'rspec'
-end
+gem 'sinatra'
+gem 'sinatra-contrib'
+
+gem 'pry', group: [:test, :development]
+gem 'rspec', group: :test
```
Comment (gotar) [inline Gemfile]: I prefer to have groups

File: config.ru [hunk 2]
```diff
@@ -0,0 +1,4 @@
+require_relative './lib/app'
+
+run App.new
+
```
Comment (gotar) [inline config.ru]: drop empty line

File: lib/app.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+require 'sinatra/reloader'
+require_relative './models/product'
+require_relative './models/warehouse_product'
+require_relative './services/store_service'
+require_relative './services/basket_service'
+require_relative './services/warehouse_service'
+require_relative './services/products_service'
+
+class App < Sinatra::Base
+
```
Comment (gotar) [inline lib/app.rb]: drop empty line

File: lib/app.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    register Sinatra::Reloader
+  end
+
+  get '/' do
+    erb :index, locals: {title: 'Home'}
+  end
+
+  get '/offer' do
+    products = @warehouse_service.products.map do |wp|
+      [@products_service.fetch(wp.product_id), wp.quantity]
```
Comment (gotar) [inline lib/app.rb]: I do it often too but better will be to use some more descriptive variable name then wp

File: lib/app.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    end
+    erb :offer, locals: {title: 'Offer', products: products}
+  end
+
+  get '/basket' do
+    products = @basket_service.products.map { |p,q| [@products_service.fetch(p), q] }
```
Comment (gotar) [inline lib/app.rb]: same here, more descriptive var name will help other devs to read your code

File: lib/services/basket_service.rb [hunk 6]
```diff
@@ -1,22 +1,22 @@
...
     unless contains?(product_id) then
       raise "Cannot remove product of id: #{product_id}, not in basket"
     end
-    @products.delete_at(@products.index { |p| p.id == product_id })
+    @products.delete_at(@products.index { |p| p == product_id })
   end
 
   def contains?(product_id)
-    @products.any? { |p| p.id == product_id }
+    @products.any? { |p| p == product_id }
```
Comment (gotar) [inline lib/services/basket_service.rb]: all those vars confuse other devs, it;s not clear you have an array of ids. maybe `product_ids` will be better

File: lib/services/products_service.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+class ProductsService
+  def initialize(products)
+    @products = products
+  end
+
+  def fetch(product_id)
+    @products.find { |p| p.id == product_id }
```
Comment (gotar) [inline lib/services/products_service.rb]: add more descriptive var name

File: spec/web_helper.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  def app
+    App
+  end
+end
+
+RSpec.configure do |config|
+  config.include Rack::Test::Methods, type: :request
+  config.include AppHelper, type: :request
+end
+
```
Comment (gotar) [inline spec/web_helper.rb]: drop empty line

File: config.ru (hunk 2 above)
Comment (filip373) [inline config.ru]: Should I add empty line on the file end? I heard that it's a good practice.

Comment (gotar) [inline config.ru]: No, not empty line, just a symbol of end of the line. Every modern editor add it automatically. If you will have problem here github will add error info at the end of file

Comment (gotar) [inline config.ru]: drop those line, and it will be ok

**Pull request merged**